modules/andes-core/broker/src/main/java/org/wso2/andes/subscription/SubscriptionBitMapHandler.java

```diff
@@ -0,0 +1,688 @@
+            }
+
+        }
+        /**
+         * Put the local subscription into the desired location
+         */
+        columnIndexOfTheSubscriptionInBitMap = localSubscriptionMapping.get(destination);
+
+        Map<String, LocalSubscription> localSubscriptionMap = localSubscriptions.get(columnIndexOfTheSubscriptionInBitMap);
+        localSubscriptionMap.put(local.getSubscriptionID(), local);
+        localSubscriptions.put(columnIndexOfTheSubscriptionInBitMap, localSubscriptionMap);
+    }
+
+    /**
+     * Method to add the ClusteredSubscription
+     *
+     * @param destination routing key of the subscription
+     * @param andes       AndesSubscription
+     *                    <p/>
+     *                    Same logic as the localSubscription
+     */
+
+    public void addClusteredSubscription(String destination,
+                                         AndesSubscription andes) {
+        int columnIndexOfTheSubscriptionInBitMap;
+
+        /**
+         * If there is no any previous subscriptions with the specified routing key
+         */
+        if (null == clusteredSubscriptionMapping.get(destination)) {
+            /**
+             * If the removed subscription places still remain take the index for the subscription
+             * from there
+             */
+            if (deletedClusters.size() > 0) {
+                columnIndexOfTheSubscriptionInBitMap = deletedClusters.remove(0);
+            } else {
+                /**
+                 * else get the index from the clusteredSubscriptionCount variable
+                 */
+                columnIndexOfTheSubscriptionInBitMap = clusteredSubscriptionCount;
+                clusteredSubscriptionCount++;
+            }
+
+            /**
+             * Put the routing key and index
+             * Mapping from the routing key to index
+             */
+            clusteredSubscriptionMapping.put(destination, columnIndexOfTheSubscriptionInBitMap);
+            /**
+             * create the entry at the index position
+             */
+            clusteredSubscriptions.put(columnIndexOfTheSubscriptionInBitMap, new ConcurrentHashMap<String, AndesSubscription>());
+
+            /**
+             * Divide the routing key into constituent parts
+             */
+            String[] destinations = destination.split("\\.");
+
+            //Loop through the constituent parts
+            for (int i = 0; i < destinations.length; i++) {
+                /**
+                 * When the constituent table is not created before
+                 * create new one Ex - If the topic is WSO2.Products.MB
+                 * But before we only had the subscriptions and topics with maximum of
+                 * 2 constituentsAdd a new table for the constituent part 3
+                 */
+                if (bitMapClustered.size() <= i) {
+                    Map<String, BitSet> newBitMapForithConstituent;
```

Comment:
Can we use a cleaner name ? e.g. bitMapForCurrentConstituent ? (will help to someone else going through the code)